26_Autoexposure/main.cpp

```diff
@@ -0,0 +1,874 @@
+					.srcSubpass = IGPURenderpass::SCreationParams::SSubpassDependency::External,
+					.dstSubpass = 0,
+					.memoryBarrier = {
+					// since we're uploading the image data we're about to draw
+					.srcStageMask = asset::PIPELINE_STAGE_FLAGS::COPY_BIT,
+					.srcAccessMask = asset::ACCESS_FLAGS::TRANSFER_WRITE_BIT,
+					.dstStageMask = asset::PIPELINE_STAGE_FLAGS::COLOR_ATTACHMENT_OUTPUT_BIT,
+					// because we clear and don't blend
+					.dstAccessMask = asset::ACCESS_FLAGS::COLOR_ATTACHMENT_WRITE_BIT
+					}
+					// leave view offsets and flags default
+				},
+				// ATTACHMENT_OPTIMAL to PRESENT_SRC
+				{
+					.srcSubpass = 0,
+					.dstSubpass = IGPURenderpass::SCreationParams::SSubpassDependency::External,
+					.memoryBarrier = {
+						.srcStageMask = asset::PIPELINE_STAGE_FLAGS::COLOR_ATTACHMENT_OUTPUT_BIT,
+						.srcAccessMask = asset::ACCESS_FLAGS::COLOR_ATTACHMENT_WRITE_BIT
+						// we can have NONE as the Destinations because the spec says so about presents
+						}
+					// leave view offsets and flags default
+				},
+				IGPURenderpass::SCreationParams::DependenciesEnd
+			};
+			scResources = std::make_unique<CDefaultSwapchainFramebuffers>(m_device.get(), swapchainParams.surfaceFormat.format, dependencies);
+			if (!scResources->getRenderpass())
+				return logFail("Failed to create Renderpass!");
+		}
+
+		// Load the shaders and create the pipelines
+		{
+			auto loadCompileAndCreateShader = [&](const std::string& relPath) -> smart_refctd_ptr<IGPUShader>
+			{
+				IAssetLoader::SAssetLoadParams lp = {};
+				lp.logger = m_logger.get();
+				lp.workingDirectory = ""; // virtual root
+				auto assetBundle = m_assetMgr->getAsset(relPath, lp);
+				const auto assets = assetBundle.getContents();
+				if (assets.empty())
+					return nullptr;
+
+				// lets go straight from ICPUSpecializedShader to IGPUSpecializedShader
+				auto source = IAsset::castDown<ICPUShader>(assets[0]);
+				if (!source)
+					return nullptr;
+				const uint32_t workgroupSize = m_physicalDevice->getLimits().maxComputeWorkGroupInvocations;
+				const uint32_t subgroupSize = m_physicalDevice->getLimits().maxSubgroupSize;
+				auto overriddenSource = CHLSLCompiler::createOverridenCopy(
+					source.get(),
+					"#define WorkgroupSize %d\n#define DeviceSubgroupSize %d\n",
+					workgroupSize,
+					subgroupSize
+				);
+
+				return m_device->createShader(overriddenSource.get());
+			};
+
+			auto createComputePipeline = [&](smart_refctd_ptr<IGPUShader>& shader, smart_refctd_ptr<IGPUComputePipeline>& pipeline, smart_refctd_ptr<IGPUPipelineLayout> pipelineLayout) -> bool
+			{
+				{
+					IGPUComputePipeline::SCreationParams params = {};
+					params.layout = pipelineLayout.get();
+					params.shader.shader = shader.get();
+					params.shader.entryPoint = "main";
+					params.shader.entries = nullptr;
+					params.shader.requireFullSubgroups = true;
+					params.shader.requiredSubgroupSize = static_cast<IGPUShader::SSpecInfo::SUBGROUP_SIZE>(5);
```

Comment:
this needs to be set from `m_physicalDevice->getLimits().maxSubgroupSize`, not just set to 32